scim/resource_entitlement.go

```diff
@@ -0,0 +1,120 @@
+	entitlementSchema := common.StructToSchema(entity{},
+		func(m map[string]*schema.Schema) map[string]*schema.Schema {
+			addEntitlementsToSchema(&m)
+			alof := []string{"group_id", "user_id", "service_principal_id"}
+			for _, field := range alof {
+				m[field].AtLeastOneOf = alof
+			}
+			return m
+		})
+	addEntitlementsToSchema(&entitlementSchema)
+	return common.Resource{
+		Create: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			return patchEntitlements(ctx, d, c, "add")
+		},
+		Read: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			split := strings.SplitN(d.Id(), "/", 2)
+			if len(split) != 2 {
+				return fmt.Errorf("ID must be two elements: %s", d.Id())
+			}
+			switch strings.ToLower(split[0]) {
+			case "group":
+				group, err := NewGroupsAPI(ctx, c).Read(split[1])
+				if err != nil {
+					return err
+				}
+				return group.Entitlements.readIntoData(d)
+			case "user":
+				user, err := NewUsersAPI(ctx, c).Read(split[1])
+				if err != nil {
+					return err
+				}
+				return user.Entitlements.readIntoData(d)
+			case "spn":
+				spn, err := NewServicePrincipalsAPI(ctx, c).Read(split[1])
+				if err != nil {
+					return err
+				}
+				return spn.Entitlements.readIntoData(d)
+			}
+			return nil
+		},
+		Update: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			return enforceEntitlement(ctx, d, c)
+		},
+		Delete: func(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+			return patchEntitlements(ctx, d, c, "remove")
+		},
+		Schema: entitlementSchema,
+	}.ToResource()
+}
+
+func patchEntitlements(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient, op string) error {
+	groupId := d.Get("group_id").(string)
+	userId := d.Get("user_id").(string)
+	spnId := d.Get("service_principal_id").(string)
+	if groupId != "" {
+		groupsAPI := NewGroupsAPI(ctx, c)
+		err := groupsAPI.UpdateEntitlements(groupId, op, readEntitlementsFromData(d))
+		if err != nil {
+			return err
+		}
+		d.SetId("group/" + groupId)
+	}
+	if userId != "" {
+		usersAPI := NewUsersAPI(ctx, c)
+		err := usersAPI.UpdateEntitlements(userId, op, readEntitlementsFromData(d))
+		if err != nil {
+			return err
+		}
+		d.SetId("user/" + userId)
+	}
+	if spnId != "" {
+		spnAPI := NewServicePrincipalsAPI(ctx, c)
+		err := spnAPI.UpdateEntitlements(spnId, op, readEntitlementsFromData(d))
+		if err != nil {
+			return err
+		}
+		d.SetId("spn/" + spnId)
+	}
+	return nil
+}
+
+func enforceEntitlement(ctx context.Context, d *schema.ResourceData, c *common.DatabricksClient) error {
+	split := strings.SplitN(d.Id(), "/", 2)
+	if len(split) != 2 {
+		return fmt.Errorf("ID must be two elements: %s", d.Id())
+	}
+	identity := strings.ToLower(split[0])
+	id := strings.ToLower(split[1])
+	switch identity {
+	case "group":
+		NewGroupsAPI(ctx, c).UpdateEntitlements(id, "remove", generateFullEntitlements())
```

Comment:
can't update and remove be combined into a single request? API expects an array of patch ops